### PR TITLE
Import CSV

### PR DESCRIPTION
**Pull request by dandaandaaaaaan**
This PR will add the feature for importing of CSV

**Sub-tasks:**
- [x] Add command
- [x] Ensure baseline file processing works
- [x] Change phone numbers to allow `+` country codes, important when importing
- [x] Skip lines if format invalid
- [x] Skip lines if duplicate
- [x] Advanced import functionality (specified params)

**Unit tests list:**
- ImportCsv model
   - [x] Constructor for null
   - [x] `isValidFilePath` method
- ImportCsv Command
   - [x] Successful import
   - [x] Successful import with certain lines that do not conform(skip lines)
   - [x] Successful import with duplicate people inside (skip lines)
   - [x] Import invalid file
   - [x] Import file empty
- ImportCsv Parser
    - [x] only file path present - > success
    - [x] file path with custom params
    - [x] file path missing -> failure
    - [x] invalid fields -> failure 
- ParserUtil
     - [x] file path valid ->success
     - [x] file path bad extension ->throw exception
     - [x] file path no extension -> throw exception
     - [x] valid position -> success
     - [x] negative position -> throw exception 
     - [x] letters in position -> throw exception
- More "advanced" features
   ~- [ ] Import Google contact (build functionality first)~ Split to next PR
   ~- [ ] Import Outlook contact (build functionality first)~ Split to next PR

Will update this as I go along in order to break down my tasks for my own management.

Issue ref: #92

### CODE REVIEW

File: src/main/java/seedu/contax/model/ImportCsv.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+package seedu.contax.model;
+
+import static seedu.contax.commons.util.AppUtil.checkArgument;
+import static seedu.contax.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.io.File;
+
+public class ImportCsv {
```
Comment (sharpstorm) [inline src/main/java/seedu/contax/model/ImportCsv.java:8]: I suggest changing the name of this class, it does not look like a noun. Use the IntelliJ refactor function.

```suggestion
public class IndexedCsvFile {
```

File: src/main/java/seedu/contax/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -132,6 +134,40 @@ public static Tag parseTag(String tag) throws ParseException {
         return tagSet;
     }
 
+    /**
+     * Parses a {@code String filePath} into a {@code Path}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code phone} is invalid.
```
Comment (sharpstorm) [inline src/main/java/seedu/contax/logic/parser/ParserUtil.java:141]: The javadoc here is probably wrong.
```suggestion
     * @throws ParseException if the given {@code filePath} is invalid.
```

File: src/main/java/seedu/contax/logic/commands/ImportCsvCommand.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            if (skippedLines.size() > 0) {
+                String skippedLinesString = "";
+                for (int i = 0; i < skippedLines.size(); i++) {
+                    skippedLinesString += skippedLines.get(i);
+                    if (i != skippedLines.size() - 1) {
+                        skippedLinesString += ", ";
+                    }
+                }
+                return new CommandResult(String.format("%s\n%s", MESSAGE_SUCCESS,
```
Comment (sharpstorm) [inline src/main/java/seedu/contax/logic/commands/ImportCsvCommand.java:94]: The parsing logical should probably be separated into another file or private function, the method is as a whole quite long.

**Review by sharpstorm — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/contax/model/ImportCsv.java (hunk 1 above)
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/model/ImportCsv.java:8]: thanks i also agree lol will change!

File: src/main/java/seedu/contax/logic/parser/ParserUtil.java (hunk 2 above)
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/logic/parser/ParserUtil.java:141]: thanks for catching!

File: src/main/java/seedu/contax/logic/commands/ImportCsvCommand.java (hunk 3 above)
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/logic/commands/ImportCsvCommand.java:94]: okay will separate to another private function

File: src/test/java/seedu/contax/logic/commands/CommandTestUtil.java [hunk 4]
```diff
@@ -94,6 +99,34 @@
...
     public static final String PREAMBLE_NON_EMPTY = "NonEmptyPreamble";
 
+    public static final String COMMAND_VALID_CSV_FILEPATH = " " + PREFIX_FILE + VALID_CSV_FILEPATH;
+    public static final String COMMAND_INVALID_BAD_EXTENSION_FILEPATH = " " + PREFIX_FILE
+            + INVALID_BAD_EXTENSION_CSV_FILEPATH;
+    public static final String COMMAND_INVALID_NO_EXTENSION_FILEPATH = " " + PREFIX_FILE
+            + INVALID_NO_EXTENSION_CSV_FILEPATH;
+    public static final String COMMAND_CSV_NAMEPOSITION = " " + PREFIX_NAME + 1;
+    public static final String COMMAND_CSV_NAMEPOSITION_CLASH = " " + PREFIX_NAME + 2;
+    public static final String COMMAND_CSV_PHONEPOSITION = " " + PREFIX_PHONE + 2;
```
Comment (HanJiyao) [inline src/test/java/seedu/contax/logic/commands/CommandTestUtil.java:109]: what would here and following +1 +2 mean, is that like magic num?

**Review by HanJiyao — APPROVED** [review]
LGTM only few minor suggestions

File: src/test/java/seedu/contax/logic/commands/CommandTestUtil.java (hunk 4 above)
Comment (dandaandaaaaaan) [reply src/test/java/seedu/contax/logic/commands/CommandTestUtil.java:109]: it's supposed to be simulated positions, so
`COMMAND_CSV_NAMEPOSITION = " " + PREFIX_NAME + 1` will be
` n/ 1` to specify name field in column 1

the clash one is set to 2 such that it clashes with another position, causing exception

I've changed the numbers to be strings instead too to make this clearer!

File: src/main/java/seedu/contax/logic/parser/ImportCsvParser.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        File file = ParserUtil.parseCsvFilePath(argMultimap.getValue(PREFIX_FILE).get());
+
+        //check if file exists, else throw exception
+
+        int namePosition = 1;
+        int phonePosition = 2;
+        int emailPosition = 3;
+        int addressPosition = 4;
+        int tagPosition = 5;
```
[the remark this answers is not in the document]
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/logic/parser/ImportCsvParser.java:41]: it isn't static variables though thats why i use normal camelcase, cos these can change if the user enters something custom

File: src/main/java/seedu/contax/model/ImportCsv.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
+package seedu.contax.model;
```
[the remark this answers is not in the document]
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/model/ImportCsv.java:1]: I put it in model as this is just the object for an Indexed File, such that it is easier to manage the imported files as an entire object. Just edited the name of the file and the dependencies as per zong han's review, so hopefully should be clearer!

File: src/main/java/seedu/contax/logic/parser/ParserUtil.java [hunk 7]
```diff
@@ -132,6 +134,40 @@ public static Tag parseTag(String tag) throws ParseException {
...
+        return new File(trimmedFilePath);
+    }
+
+    /**
+     * Parses a {@String position} into an {@code integer}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException
+     */
+    public static int parseCsvPositions(String position) throws ParseException {
```
[the remark this answers is not in the document]
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/logic/parser/ParserUtil.java:158]: thanks for the suggestion! will add

File: src/main/java/seedu/contax/model/person/Phone.java [hunk 8]
```diff
@@ -11,8 +11,9 @@
 
 
     public static final String MESSAGE_CONSTRAINTS =
-            "Phone numbers should only contain numbers, and it should be at least 3 digits long";
-    public static final String VALIDATION_REGEX = "\\d{3,}";
+            "Phone numbers should only contain numbers and country code (optional), and it should be "
```
[the remark this answers is not in the document]
Comment (dandaandaaaaaan) [reply src/main/java/seedu/contax/model/person/Phone.java:14]: thanks for the suggestion! changing

File: src/test/java/seedu/contax/logic/commands/CommandTestUtil.java [hunk 9]
```diff
@@ -94,6 +99,34 @@
...
+    public static final String COMMAND_CSV_INVALID_PHONEPOSITION = " " + PREFIX_PHONE + "text";
+    public static final String COMMAND_CSV_INVALID_EMAILPOSITION = " " + PREFIX_EMAIL + "text";
+    public static final String COMMAND_CSV_INVALID_ADDRESSPOSITION = " " + PREFIX_ADDRESS + "text";
+    public static final String COMMAND_CSV_INVALID_TAGPOSITION = " " + PREFIX_TAG + "text";
+
+    public static final String COMMAND_CSV_NEGATIVE_NAMEPOSITION = " " + PREFIX_NAME + -1;
+    public static final String COMMAND_CSV_NEGATIVE_PHONEPOSITION = " " + PREFIX_PHONE + -1;
+    public static final String COMMAND_CSV_NEGATIVE_EMAILPOSITION = " " + PREFIX_EMAIL + -1;
+    public static final String COMMAND_CSV_NEGATIVE_ADDRESSPOSITION = " " + PREFIX_ADDRESS + -1;
+    public static final String COMMAND_CSV_NEGATIVE_TAGPOSITION = " " + PREFIX_TAG + -1;
```
[the remark this answers is not in the document]
Comment (dandaandaaaaaan) [reply src/test/java/seedu/contax/logic/commands/CommandTestUtil.java]: thanks for the spot! changed

**Review by sharpstorm — APPROVED** [review]
LGTM.

**Pull request merged**